community/kernel/src/main/java/org/neo4j/storageengine/api/schema/SchemaRule.java

```diff
@@ -26,23 +26,97 @@
 import org.neo4j.kernel.api.schema_new.SchemaDescriptor;
 import org.neo4j.kernel.api.schema_new.constaints.ConstraintDescriptor;
 import org.neo4j.kernel.api.schema_new.index.NewIndexDescriptor;
+import org.neo4j.kernel.impl.store.record.ConstraintRule;
+import org.neo4j.kernel.impl.store.record.IndexRule;
 import org.neo4j.kernel.impl.store.record.RecordSerializable;
 
 /**
  * Represents a stored schema rule.
  */
-public interface SchemaRule extends SchemaDescriptor.Supplier, RecordSerializable
+public abstract class SchemaRule implements SchemaDescriptor.Supplier, RecordSerializable
 {
+    protected final long id;
+    protected String name;
+
+    protected SchemaRule( long id )
+    {
+        this.id = id;
+    }
+
     /**
      * The persistence id for this rule.
      */
-    long getId();
+    public final long getId()
+    {
+        return this.id;
+    }
+
+    /**
+     * @return The (possibly user supplied) name of this schema rule.
+     */
+    public final String getName()
+    {
+        if ( name == null )
+        {
+            name = SchemaRule.generateName( this );
+        }
+        return name;
+    }
+
+    /**
+     * Set the name of this rule, if it has not been set already.
+     *
+     * Note that once a name has been observed - either via this method or via {@link #getName()} – for this rule,
+     * it cannot be changed.
+     * @param name The name desired for this rule.
+     * @return {@code true} if the name was set, {@code false} if this rule already has a name.
+     */
```

Comment:
I don't know, I find this to be a bit too complicated? How about making the name a final field instead, and pass it into the constructor of the SchemaRules. Then we can make a factory method without the name, which would auto-generate the name. This would also avoid the instance of checks in `generateName`.